src/chatd.cpp

```diff
@@ -4315,11 +4471,116 @@ void FilteredHistory::clear()
     init();
 }
 
+HistSource FilteredHistory::loadHistory(uint32_t count)
+{
+    uint32_t messagesLoaded = 0;
+    if ((mOldestNotifyMsg != --mBuffer.end() && mBuffer.size() > 1) || mFirstNotification)
+    {
+        mFirstNotification = false;
+        for (auto it = mOldestNotifyMsg; it != mBuffer.end() && messagesLoaded < count; it++)
+        {
+            Idx index = mNewest - std::distance(mBuffer.begin(), it);
+            CALL_LISTENER_FH(onLoaded,  &(*(*it)), index);
+            messagesLoaded++;
+            mOldestNotifyMsg = it;
+        }
+
+        if (messagesLoaded)
+        {
+            return HistSource::kHistSourceRam;
+        }
+    }
+
+    uint32_t messagesLoadedFromdb = loadHistoryFromDb(count);
+    mOldestNotifyMsg = --mBuffer.end();
+    if (messagesLoadedFromdb)
+    {
+        mFirstNotification = false;
+        return HistSource::kHistSourceDb;
+    }
+
+    if (!hasAllHistory() && mChat->connection().isOnline())
```

Comment:
Here you probably want to check that the connection is actually logged-in into chatd, not just the connection is opened. Note that the connection may be online, but the chat may not be logged-in yet.
I'll change that. Let me know if my assumption is wrong.